examples/quantum_volume.py

```diff
@@ -0,0 +1,153 @@
+                           rs=None) -> cirq.Circuit:
+    """Generates a model circuit with the given number of qubits and depth.
+
+    The generated circuit consists of `depth` layers of random qubit
+    permutations followed by random two-qubit gates that are sampled from the
+    Haar measure on SU(r).
+
+    Args:
+        num_qubits: The number of qubits in the generated circuit.
+        depth: The number of layers in the circuit.
+        rs: A numeric or RandomState to seed the qubit permutations with.
+
+    Returns:
+      The generated circuit.
+    """
+    # Setup the circuit and its qubits.
+    qubits = cirq.LineQubit.range(num_qubits)
+    circuit = cirq.Circuit()
+    if not isinstance(rs, np.random.RandomState):
+        rs = np.random.RandomState(rs)
+
+    # For each layer.
+    for _ in range(depth):
+        # Generate uniformly random permutation Pj of [0...n-1]
+        perm = rs.permutation(num_qubits)
+
+        # For each consecutive pair in Pj, generate Haar random SU(4)
+        # Decompose each SU(4) into CNOT + SU(2) and add to Ci
+        for k in range(math.floor(num_qubits / 2)):
+            permuted_indices = [int(perm[2 * k]), int(perm[2 * k + 1])]
+            special_unitary = cirq.testing.random_special_unitary(4)
+
+            # Convert the decomposed unitary to Cirq operations and add them to
+            # the circuit.
+            ops = cirq.two_qubit_matrix_to_operations(
+                qubits[permuted_indices[0]],
+                qubits[permuted_indices[1]],
+                special_unitary,
+                allow_partial_czs=False)
+            circuit.append(ops)
+
+    # Don't measure all of the qubits at the end of the circuit because we will
+    # need to classically simulate it to compute its heavy set.
+    return circuit
+
+
+def compute_heavy_set(circuit: cirq.Circuit) -> List[str]:
+    """Classically compute the heavy set of the given circuit.
+
+    The heavy set is defined as the output bit-strings that have a greater than
+    median probability of being generated.
+
+    Args:
+        circuit: The circuit to classically simulate.
+
+    Returns:
+        A list containing all of the heavy bit-string results.
+    """
+    # Classically compute the probabilities of each output bit-string through
+    # simulation.
+    simulator = cirq.Simulator()
+    results = cast(cirq.WaveFunctionTrialResult,
+                   simulator.simulate(program=circuit))
+    # Compute the median probability of the output bit-strings.
+    median = statistics.median(results.state_vector())
+
+    # The output wave function is a vector from the result value (big-endian) to
+    # the probability of that bit-string. Compute a format string converts the
+    # given index to the corresponding qubit string.
+    format_str = '{0:0%sb}' % len(circuit.all_qubits())
+    # Return all of the bit-strings that have a probability greater than the
+    # median.
+    return ([
```

Comment:
Is the quantum computer allowed to count the same bit string multiple times in this test?